### PR TITLE
Issue 38148: Allow sample parents to be completely removed

### PR DESCRIPTION
**Pull request by labkey-tchad**
[Related issue](https://www.labkey.org/home/Developer/issues/issues-update.view?issueId=40083)

### CODE REVIEW

**Review by labkey-susanh — COMMENTED** [review]
Tests look good to me.  I defer to test team for code structure review.

File: src/org/labkey/test/params/experiment/DataClassDefinition.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+public class DataClassDefinition extends DomainProps
+{
+    private String _name;
+    private String _nameExpression;
+    private String _description;
+    private String _category;
+    private String _materialSource;
+    private List<FieldDefinition> _fields = new ArrayList<>();
+
+    public DataClassDefinition() { }
```
Comment (labkey-danield) [inline src/org/labkey/test/params/experiment/DataClassDefinition.java]: Are you sure you want to keep this constructor? Having this constructor, and a setter for the name property will allow _name to be null which is never checked in getOptions.

File: src/org/labkey/test/params/experiment/SampleSetDefinition.java [hunk 2]
```diff
@@ -170,4 +172,16 @@ protected String getKind()
         }
         return options;
     }
+
+    @Override
+    protected @NotNull String getSchemaName()
```
Comment (labkey-danield) [inline src/org/labkey/test/params/experiment/SampleSetDefinition.java:175]: Odd place for the @NotNull annotation.

File: src/org/labkey/test/params/list/ListDefinition.java [hunk 3]
```diff
@@ -122,4 +122,16 @@ protected String getKind()
         options.put("keyName", getKeyName());
         return options;
     }
+
+    @Override
+    protected @NotNull String getSchemaName()
```
Comment (labkey-danield) [inline src/org/labkey/test/params/list/ListDefinition.java:127]: Similar odd place for @NotNull.

File: src/org/labkey/test/tests/SampleSetTest.java [hunk 4]
```diff
@@ -620,6 +611,109 @@ public void deleteLineageParent() throws IOException, CommandException
...
+        final String sampleI = "DPS-I";
+        sampleGenerator.addCustomRow(Map.of("name", sampleC, "strCol", "c-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleD, "strCol", "d-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleE, "strCol", "e-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleF, "strCol", "f-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleG, "strCol", "g-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleH, "strCol", "h-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        sampleGenerator.addCustomRow(Map.of("name", sampleI, "strCol", "i-v1", sampleParentKey, sampleParents, dataParentKey, dataParents));
+        SaveRowsResponse saveRowsResponse = sampleGenerator.insertRows(createDefaultConnection(true), sampleGenerator.getRows());
```
Comment (labkey-danield) [inline src/org/labkey/test/tests/SampleSetTest.java]: This return value is never used.

**Review by labkey-danield — APPROVED** [review]
Looks ok to me. Even though the changes to TestDataGenerator are small you may want to add @labkey-chrisj to the review.

File: src/org/labkey/test/params/experiment/SampleSetDefinition.java (hunk 2 above)
Comment (labkey-tchad) [reply src/org/labkey/test/params/experiment/SampleSetDefinition.java:175]: That's IntelliJ's template for overriding a method. I hadn't notice that before. I guess it makes some kind of sense to have `NotNull` close to the method's return type.

Comment (labkey-adam) [reply src/org/labkey/test/params/experiment/SampleSetDefinition.java:175]: FWIW, that's my preferred location for ```@NotNull``` on the return type

**Pull request merged**